[Wise transfer] Temporarily disable for maintenance

garyhtou opened the pull request:
## Summary of the problem

At the moment, operations is overwhelmed with the number of Wise transfers, and there are a few serious bugs that are making it even harder for them. We've [decided to temporarily disable Wise transfers](https://hackclub.slack.com/archives/C026RKHLPNJ/p1759263275754389?thread_ts=1759261355.207749&cid=C026RKHLPNJ) until we get these bugs fixed up.


This PR disables Wise transfers for both reimbursements (as a payout method) and also for event transfers.

## Describe your changes

I've decided to add more tooling around disabling payout methods. It's typically a pain to disable a payout method since there are so many locations you need to add checks and messaging. Now, you can simply update the `User::PayoutMethod::UNSUPPORTED_METHODS` constant to declaratively disable a payout method.

`User::PayoutMethod::UNSUPPORTED_METHODS` does not affect the transfer options that events have access to. The process for disabling that is still fairly manual, although it's [not as bad](https://github.com/hackclub/hcb/commit/275b46906a7587633081743084e4211002dde98e) as reimbursement payout methods.

garyhtou committed:
Add the concept of unsupported payout methods

garyhtou committed:
Oh! apparently Wire and WiseTransfer didn't include the `Shared` concern

I think they should?

garyhtou committed:
Convert `!= PaypalTransfer` check to use `unsupported?`

garyhtou committed:
Dry up user payout method validation

and add more useful error message (including the reason why this method is unsupported)

garyhtou committed:
Dry up payout form radio box options to allow for declarative disabling of payout methods

garyhtou committed:
Remove redundant validation

Handled by https://github.com/hackclub/hcb/blob/ef21364852df6b5a07d749e26ab9a3ea7fa589b7/app/models/user/payout_method/shared.rb#L23-L27

garyhtou committed:
WIP controller test

garyhtou committed:
Disable wise transfer for events

garyhtou committed:
Lint

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

davidcornu commented on app/models/reimbursement/report.rb:
```diff
@@ -109,8 +109,8 @@ class Report < ApplicationRecord
           guard do
             user.payout_method.present? && event && !exceeds_maximum_amount? && !below_minimum_amount? &&
-              expenses.any? && !missing_receipts? && user.payout_method.class != User::PayoutMethod::PaypalTransfer &&
-              !event.financially_frozen? && expenses.none? { |e| e.amount.zero? } && !mismatched_currency?
+              expenses.any? && !missing_receipts? && !event.financially_frozen? && expenses.none? { |e| e.amount.zero? } &&
+              !mismatched_currency? && payout_method_allowed?
```
- **concern**: This `guard` block is really getting out of hand. Might I suggest we at least split the conditions into separate lines
- **note**: I couldn't make heads or tails of this diff so here's my manual audit: 
	Predicate | Change
	---- | ----
	`user.payout_method.present?` | Kept
	`event` | Kept
	`!exceeds_maximum_amount?` | Kept
	`!below_minimum_amount?` | Kept
	`expenses.any?` | Kept
	`!missing_receipts?` | Kept
	`user.payout_method.class != User::PayoutMethod::PaypalTransfer` | Replaced with `payment_method_allowed?`
	`!event.financially_frozen?` | Kept
	`expenses.none? { \|e\| e.amount.zero? }` | Kept
	`!mismatched_currency?` | Kept

garyhtou replied:
Yeah, this definitely is getting out of hand. I suggest we tackle that in another PR?

davidcornu committed:
Update spec to match controller behaviour

- The error message is grabbed off of `user.payout_method` directly so
  won't have the "Payout method is invalid prefix"
- Updated the expectation to just check `flash` directly
- Put the string in a `reason` variable to make it clearer that it
  should be the same. Also appended "in tests" at the end to make sure
  our stubbed constant is actually having an effect on the outcome.

davidcornu committed:
Render all `payout_method` errors

This doesn't fix known bugs as far as I know but it looks suspicious enough that I ended
up looking into it as a potential issue. To spare future readers, let's
just make sure we don't skip errors here.

davidcornu committed:
Get `ReimbursementMailer` spec working again

garyhtou committed:
Use `Hash#key?` instead of `Object#in?(Hash#keys)`

Prevent us from allocating an array

Co-authored-by: David Cornu <davidc@hackclub.com>

garyhtou commented on app/models/user/payout_method/shared.rb:
```diff
@@ -19,6 +19,25 @@ module Shared
+
+      class_methods do
+        def unsupported?
+          self.in?(User::PayoutMethod::UNSUPPORTED_METHODS.keys)
```
Great suggestion! [`be8a4b7` (#11770)](https://github.com/hackclub/hcb/pull/11770/commits/be8a4b7cab223d4c224b2849ccf8d71530d0916a)

garyhtou commented on app/models/user.rb:
```diff
@@ -525,8 +525,15 @@ def on_phone_number_update
-    unless payout_method_type.nil? || payout_method.is_a?(User::PayoutMethod::Check) || payout_method.is_a?(User::PayoutMethod::AchTransfer) || payout_method.is_a?(User::PayoutMethod::PaypalTransfer) || payout_method.is_a?(User::PayoutMethod::Wire) || payout_method.is_a?(User::PayoutMethod::WiseTransfer)
-      errors.add(:payout_method, "is an invalid method, must be check, PayPal, wire, Wise transfer, or ACH transfer")
+    if payout_method_type.present? && User::PayoutMethod::SUPPORTED_METHODS.none? { |method| payout_method.is_a?(method) }
+      # I'm using `try` here in the slim chance that `payout_method` is some
+      # random model and doesn't include `User::PayoutMethod::Shared`.
```
I agree, will tackle in another PR

garyhtou committed:
Move inline CSS to tailwind

Co-authored-by: David Cornu <davidc@hackclub.com>

garyhtou commented on app/views/reimbursement/reports/_actions.html.erb:
```diff
@@ -86,7 +86,7 @@
 
       <div style="flex-grow: 1;" class="flex justify-end">
-        <% if !user.payout_method.present? || report.payout_holding&.failed? || (user.payout_method.present? && user.payout_method.class == User::PayoutMethod::PaypalTransfer) || report.mismatched_currency? %>
+        <% if !user.payout_method.present? || report.payout_holding&.failed? || (user.payout_method.present? && user.payout_method.unsupported?) || report.mismatched_currency? %>
```
Yeah, will create a ticket for this

garyhtou committed:
Merge branch 'main' into garyhtou/disable-wise-transfers
